### DISCUSSION TITLE
See Cython and C++ compiler console output in editable install

### DISCUSSION THREAD

**Discussion by joshayers** [mesonbuild/meson-python#546] [Q&A] [answered]
I have a project that uses Cython and C++.  Is it possible to see the console output from Cython and the C++ compiler?  It often contains helpful warning messages.  It's not shown in the console (even in verbose mode) and I couldn't find it in any of the logs in the build directory.

For reference, I removed the types from a loop variable in a Cython file.  I know Cython issues a warning in that case, but this image is all the output that is shown.

<img width="874" alt="console" src="https://github.com/mesonbuild/meson-python/assets/831565/24922b2a-2c88-48fd-8ea3-b20e592daed9">

This is similar to [discussion 505](https://github.com/mesonbuild/meson-python/discussions/505).  In that case, the person was looking for the compiler flags, while I would like the output.

Thanks in advance for any help!

**Comment by rgommers** [comment]
Hi @joshayers, the actual compilation is done by `ninja`, so you want to execute that in verbose mode. To do so, you need the `-Ccompile-args=-v` argument, e.g. to build a wheel in verbose mode with `build`:
```
python -m build -wnx -Ccompile-args=-v
```

This should show you output like this:
```
[304/305] arm64-apple-darwin20.0.0-clang -Inumpy/random/_generator.cpython-39-darwin.so.p -Inumpy/random -I../numpy/random -I../numpy/random/src -Inumpy/_core -I../numpy/_core -Inumpy/_core/include -I../numpy/_core/include -I../numpy/_core/src/common -Inumpy -I/Users/rgommers/mambaforge/envs/numpy-dev/include/python3.9 -I/Users/rgommers/code/numpy/.mesonpy-t8cm41h1/meson_cpu -fvisibility=hidden -fcolor-diagnostics -DNDEBUG -w -std=c99 -O3 -fno-strict-aliasing -ftrapping-math -Wno-unsupported-floating-point-opt -DNPY_HAVE_NEON_VFPV4 -DNPY_HAVE_NEON_FP16 -DNPY_HAVE_NEON -DNPY_HAVE_ASIMD -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -isystem /Users/rgommers/mambaforge/envs/numpy-dev/include -D_FORTIFY_SOURCE=2 -isystem /Users/rgommers/mambaforge/envs/numpy-dev/include -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -DNPY_NO_DEPRECATED_API=0 -MD -MQ numpy/random/_generator.cpython-39-darwin.so.p/meson-generated_numpy_random__generator.pyx.c.o -MF numpy/random/_generator.cpython-39-darwin.so.p/meson-generated_numpy_random__generator.pyx.c.o.d -o numpy/random/_generator.cpython-39-darwin.so.p/meson-generated_numpy_random__generator.pyx.c.o -c numpy/random/_generator.cpython-39-darwin.so.p/numpy/random/_generator.pyx.c
[305/305] arm64-apple-darwin20.0.0-clang++  -o numpy/random/_generator.cpython-39-darwin.so numpy/random/_generator.cpython-39-darwin.so.p/meson-generated_numpy_random__generator.pyx.c.o -L/Users/rgommers/mambaforge/envs/numpy-dev/lib -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -Wl,-undefined,dynamic_lookup -bundle -Wl,-undefined,dynamic_lookup -Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,/Users/rgommers/mambaforge/envs/numpy-dev/lib -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /Users/rgommers/mambaforge/envs/numpy-dev/include -D_FORTIFY_SOURCE=2 -isystem /Users/rgommers/mambaforge/envs/numpy-dev/include numpy/random/libnpyrandom.a numpy/_core/libnpymath.a
ld: warning: -pie being ignored. It is only used when linking a main executable
```

**Reply by joshayers** [reply]
Hi @rgommers, thanks for the reply.

Unfortunately, I don't think that works.  That prints the compiler input - the full command line with all the arguments - but not any warnings or other output from the compiler.

I attached two examples from compiling the same code, once with setuptools and once with meson.  The code mixes signed and unsigned integers in a comparison (see below), which causes a compiler warning.  It doesn't appear in the meson output.

<img width="199" alt="signed-unsigned" src="https://github.com/mesonbuild/meson-python/assets/831565/0f93b5b9-1339-421f-a1eb-2877fb7a5c9b">

[setuptools_output_msvc.txt](https://github.com/mesonbuild/meson-python/files/13503519/setuptools_output.txt)
[meson_output_msvc.txt](https://github.com/mesonbuild/meson-python/files/13503523/meson_output.txt)

To double check, I ran the same thing in gcc.  Differently worded warning, but same result.

[setuptools_output_gcc.txt](https://github.com/mesonbuild/meson-python/files/13503607/setuptools_output_gcc.txt)
[meson_output_gcc.txt](https://github.com/mesonbuild/meson-python/files/13503608/meson_output_gcc.txt)

Is this more of a `ninja` issue, since it performs the compilation?  I wonder if it's just capturing `stdout` but not `stderr`.

Thanks again for the reply.

**Reply by rgommers** [reply]
Compiler warnings should always be shown, even in non-verbose mode. I have never seen a meson-python build _not_ show them I think. Unless there's a difference between editable and regular builds here? Do you see the warning with a regular `python -m build` or `pip install .`?

**Reply by joshayers** [reply]
No, I don't see the warnings with either of those commands.  When I build with setuptools I see them, but not with meson.

Let me know if there's any other info that would help diagnose.

**Reply by joshayers** [reply]
I tried running ninja directly, using the command that's created by meson.  Here's the only output I get, even with the `-v` flag.  No compiler warnings.  Note that I ran it twice, first time without the `-v` flag and the second time with it.

<img width="666" alt="ninja output" src="https://github.com/mesonbuild/meson-python/assets/831565/00c7abdb-5b5a-4e3e-ae0b-36d84cd9692d">

When the compilation fails completely, those errors are shown, but warnings are being swallowed for some reason.  It seems to be an issue in ninja, so I'll investigate there further.  If you have any suggestions, that'd be appreciated.  Thanks again!

**Reply by joshayers** [reply]
I found the problem.  The `build.ninja` file included "-w" as an argument for all the compiler calls.  That disables all warnings.  I replaced it with "/Wall" and I get an entire screen-full of warnings.

I tried setting a different warning level in my top level `meson.build` file but it didn't change anything.  What is the correct way to change this warning level?

    project(
      'rs_example_lib', 'cython', 'cpp',
      license: 'Apache-2.0 or LGPL-2.1-or-later',
      meson_version: '>= 1.3.0',
      default_options: ['warning_level=2'],
    )`

**Reply by rgommers** [reply]
Ah, that makes sense. Did you have it set to `warning_level=0` explicitly then? The default is `1`, and that has always worked well enough that I never found a reason to adjust it.

**Reply by joshayers** [reply]
No, I didn't have it set at all.  Setting the warning level in the `project` section doesn't change anything.  The `-w` suppress all warnings argument is still there.

I tried this to override it:

    project(
      'rs_example_lib', 'cython', 'cpp',
      license: 'Apache-2.0 or LGPL-2.1-or-later',
      meson_version: '>= 1.3.0',
    )
    
    if meson.get_compiler('cpp').get_id() == 'msvc'
      warn_arg = '/W3'
    else
      warn_arg = '-Wall'
    endif
    add_global_arguments(warn_arg, language: 'cpp')

It correctly adds the additional argument in the `build.ninja` file, but `-w` is still there.  It works in MSVC (`/W3` takes precedence over `-w`) but with gcc `-w` seems to take precedence over `-Wall` and warnings are still suppressed.

Here are the compiler command line arguments copied from `build.ninja`.  Where do the default compiler args come from?  Is there any method in `meson.build` to explicitly remove them?

MSVC:
`ARGS = "-Irs_example_lib\frf.cp311-win_amd64.pyd.p" "-Irs_example_lib" "-I..\..\rs_example_lib" "-I..\..\.venv\Lib\site-packages\numpy\core\include" "-IC:\Program$ Files\Python311\Include" "-DNDEBUG" "/MD" "/nologo" "/showIncludes" "/utf-8" "/Zc:__cplusplus" "-w" "/EHsc" "/O2" "/Gw" "/W3" "-DMS_WIN64=" "-DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION" "/Fdrs_example_lib\frf.cp311-win_amd64.pyd.p\meson-generated_rs_example_lib_frf.pyx.cpp.pdb"`

GCC:
`ARGS = -Irs_example_lib/frf.cpython-311-x86_64-linux-gnu.so.p -Irs_example_lib -I../../rs_example_lib -I../../.venv/lib/python3.11/site-packages/numpy/core/include -I/usr/include/python3.11 -I/usr/include/x86_64-linux-gnu/python3.11 -fvisibility=hidden -fvisibility-inlines-hidden -fdiagnostics-color=always -DNDEBUG -D_FILE_OFFSET_BITS=64 -w -O3 -Wall -fPIC -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION`

**Comment by dnicolodi** [comment]
I suspect that the `-w` comes from the compilation flags injected by one of the dependencies of your project. Without having access to a reproducer for the problem, I don't think we cannot help you further.

**Reply by joshayers** [reply]
@dnicolodi @rgommers Here's a minimal example that reproduces the problem.  The only dependencies are NumPy, Cython, build, and Meson.

https://github.com/joshayers/meson_test

I tested it using Windows/MSVC and Linux/gcc.  The build.ninja file included the "-w" flag that suppresses all compiler warnings in both cases.

Thanks again for your help.  Let me know if you're not able to reproduce the problem or need more info.

**Reply by dnicolodi** [reply]
Turns out that Meson disables warnings when compiling the C or C++ code generated by Cython: https://github.com/mesonbuild/meson/blob/30184a48a032e6a2a6dbf0a76e437aa676c3e4aa/mesonbuild/backend/ninjabackend.py#L2830-L2832 The conditional there checks for Vala generated source code, but the Cython support is bolted on the side of the Vala support and the code does not distinguish which transpiler emitted the C code.

I am not sure whether getting the C compiler warnings for generated code makes sense or not.

**Reply by joshayers** [reply]
Cython produces a minimal number of warnings in my experience, at least at the default warning level.

A common use case for it is wrapping C or C++ code.  I've always done that by compiling the Cython-generated code and the manually-written C or C++ into the same module.  Here's one example of the same thing in NumPy: https://github.com/numpy/numpy/blob/023c65180cccd53108e1f2dc19fd44dd397ff655/numpy/random/meson.build#L59.  It seems undesirable to hide warnings in that case.

Would it be possible to add a config option for this?  Or detect that the compiler is Cython instead of Vala and not disable warnings in that case?

Thanks again.

**Reply by dnicolodi** [reply]
By the way, your test project specifies `meson-python`, `meson`, `cython`, and `build` as runtime dependencies. Most likely, you don't want to do that. The first three are only build time dependencies. `build` should not be listed as a dependency, not even as a build time one: users could use a different Python package build front-end, such as `pip`.

Also, you explicitly add the Python (`dep_py = py.dependency()`) a as dependency for the Python extension modules. This is not necessary since at least Meson 0.63.

**Comment by dnicolodi** [comment]
> A common use case for it is wrapping C or C++ code. I've always done that by compiling the Cython-generated code and the manually-written C or C++ into the same module. Here's one example of the same thing in NumPy: https://github.com/numpy/numpy/blob/023c65180cccd53108e1f2dc19fd44dd397ff655/numpy/random/meson.build#L59. It seems undesirable to hide warnings in that case.

This use case is not affected. The warnings are suppressed only in the compilation of the generated C or C++ source code, not during the compilation of all the source code involved in producing a build artifact.

> Would it be possible to add a config option for this?

I don't know how such an option would look like. Meson does not expose any facility to set compilation options per source file.

>Or detect that the compiler is Cython instead of Vala and not disable warnings in that case?

I thought that this could be the solution. However, this would not be a backward compatible change: project that use `-Werror` would start to fail to compile if their Cython code emits a warning. In general, users do not have a way to resolve warnings emitted by the C or C++ compiler when compiling generated code. Therefore, it is of little use to emit warnings for that code, as they cannot be acted upon. As explained, emitting warnings for generated code is not compatible with using `-Werror` to catch warnings in parts of the code base on which the user has direct control.

@dcbaker @eli-schwartz opinions?

**Reply by joshayers** [reply]
Using `-Werror` has no effect when building a Cython module with gcc.  It is overridden by `-w` and any warnings are suppressed.

Steps to reproduce, using the [meson_test](https://github.com/joshayers/meson_test) repo from my earlier comment.

1. Edit top-level `meson.build` to add `-Werror` as a compiler flag.

        add_global_arguments('-Werror', language: 'cpp')

2. Set up an editable install as described in the meson_test  repo's readme.
3. Edit rs_example_lib/rs_example_lib/trapz.hpp to add a signed/unsigned comparison.  Note that this is code that I wrote - it is not Cython-generated.  It is used in the `trapz.pyx` module via the Cython `extern` statement.

        static inline double
        trapz_lib(double x_min, double x_max, double const *y, npy_int64 length)
        {
            const npy_uint64 n = length - 1; // <== change this to an unsigned integer
            double accum = y[0];
            #pragma omp parallel for reduction(+ : accum) schedule(guided)
            for (npy_int64 i = 1; i < n; i++) // <== signed/unsigned comparison here
            {
                accum += 2.0 * y[i];
            }
            accum += y[n];
            return 0.5 * (x_max - x_min) / static_cast<double>(n) * accum;
        }
4. Touch rs_example_lib/rs_example_lib/trapz.hpp to trigger a rebuild.
5. Run Python and import `rs_example_lib.trapz`.
6. No errors or warnings are emitted.

<img width="807" alt="with -w flag" src="https://github.com/mesonbuild/meson-python/assets/831565/221defaa-3280-4598-a8ed-0de9144ad15b">

7. Manually edit `build/cp31n/build.ninja` to remove the `-w` flag from compiler args for `trapz.pyx` module.
8. Touch rs_example_lib/rs_example_lib/trapz.hpp to trigger a rebuild.
9. Run Python and import `rs_example_lib.trapz`.
10. Compilation fails due to signed/unsigned comparison

<img width="843" alt="without -w flag" src="https://github.com/mesonbuild/meson-python/assets/831565/c99ed082-5bdd-48c3-b9d9-ea68d7881e38">

This is very unexpected behavior IMO.  At a minimum, I think it should be explained in the docs.  Would you accept a PR that updates the Cython section to explain that warnings are always suppressed (and it's not possible to override)?

**Reply by dnicolodi** [reply]
> Using `-Werror` has no effect when building a Cython module with gcc. It is overridden by `-w` and any warnings are suppressed.

Who said that it does *now*? It will if the current behavior is changed.

**Reply by eli-schwartz** [reply]
My opinion is that it's unexpected to use `-w` for vala, whether it affects cython as a side effect or not.

**Reply by eli-schwartz** [reply]
@thesamesam

Behold, a horror occurs. Remember how we got all upset about lxml...

Not to mention vala is utterly broken with Modern C and compilers are about to get real mad by default. So the foundational idea behind this was broken.

**Reply by dcbaker** [reply]
How many cython users do we have, realistically? I think that number might be small enough that even if it's a breaking change in theory as long as it doesn't break *those* projects I would be fine with it.

It seems like something that I accidentally inherited from vala when I generalized that code rather than something that I did on purpose...

**Reply by eli-schwartz** [reply]
It's even more important to stop using -w in Vala targets than it is to stop using it in cython targets.

The number of extremely dangerous UB issues in Vala that are going unreported is much higher than in cython. It is not meson's job to lie to users and pretend that vala is a safe programming language (by any definition of language safety) when compilers are defaulting to telling you that the generated code violates all the semantic laws of the C programming language.

It is harmful to users.

**Reply by dnicolodi** [reply]
https://github.com/mesonbuild/meson/pull/12596 Is a minimally invasive fix. It removes the warning suppression for Cython generated source code, while keeping it for Vala. I don't know enough about Vala to say whether suppressing the warnings is a good idea or not.

**Reply by eli-schwartz** [reply]
I'm already working on this via https://github.com/eli-schwartz/meson/commit/2e8bb87508f490ff2e90997a1a669ed9fb3ebb5b but I missed reverting a bit.

**Reply by eli-schwartz** [reply, marked as answer]
https://github.com/mesonbuild/meson/pull/12597

**Reply by rgommers** [reply]
Thanks for fixing this!

Now that the fun is over, I realized why I do see Cython warnings in SciPy: we're always using generators to invoke `cython`, and then pass the resulting C/C++ files as regular source targets to `py.extension_module`, rather than passing that the `.pyx` sources. So we sidestepped this `-w` issue by accident.

> How many cython users do we have, realistically? I think that number might be small enough that even if it's a breaking change in theory as long as it doesn't break those projects I would be fine with it.

Hard to say, but growing at a steady pace. Many meson-python users use Cython, and we crossed the 0.5M downloads/month threshold. I've seen ~15 significant packages adopt it, and more smaller/personal projects.
